### ISSUE TITLE
Problem playing .bik movie files ingame while using d3d8to9

### ISSUE THREAD

**Issue by Radio30** [crosire/d3d8to9#61]
This in Indiana Jones and the Emporer's Tomb.

All the intro and cut-scene movies are in .bik format and while using d3d8to9 the audio plays during the videos but the screen is black.
Any idea what could be causing this?

**Comment by elishacloud** [comment]
Yes this issue was introduced with Windows 10 [Fall Creators Edition](https://blogs.windows.com/windowsexperience/2017/10/17/whats-new-windows-10-fall-creators-update/).  If you are running any older version of Windows 10 (or other versions of Windows) this issue does not happen.  I noticed this a while ago and was planning to file an issue on this, but you saved me the trouble.

I have seen this "black screen" issue on a number of games starting with the [Creators Edition](https://blogs.windows.com/windowsexperience/2017/04/11/whats-new-in-the-windows-10-creators-update/#tLCYX7W76LVRXEEE.97).  See [here](https://github.com/narzoul/DDrawCompat/issues/15) for a similar example.

**Comment by HerMajestyDrMona** [comment]
Just information: I'm using Windows 10 Fall Creators Update (16299.248 currently) and .bik movies work fine in Empire Earth 2. They work with and without d3d8to9.

**Comment by elishacloud** [comment]
Yes, I believe the issue is specific to _Indiana Jones and the Emporer's Tomb_ and not directly related to .bik movies.

**Comment by wetcoder** [comment]
**[Indiana Jones and The Emperor's Tomb (2003) Full controller support FIX  + Widescreen Fix](https://www.gog.com/forum/indiana_jones_series/controller_support_for_emperors_tomb/post7)**

**Comment by elishacloud** [comment]
I just found out that _Indiana Jones and the Emperor's Tomb_ uses [DirectDraw](https://msdn.microsoft.com/en-us/library/windows/desktop/gg426115(v=vs.85).aspx).  I noticed it calling:

- [DirectDrawCreate](https://msdn.microsoft.com/en-us/library/windows/desktop/gg426116(v=vs.85).aspx)
- [SetCooperativeLevel](https://msdn.microsoft.com/en-us/library/windows/desktop/gg426155(v=vs.85).aspx)
- [CreateSurface](https://msdn.microsoft.com/en-us/library/windows/desktop/gg426136(v=vs.85).aspx)

However `CreateSurface` fails with error 887600E1.  This happens with or without d3d8to9.

The game tries to use both DirectDraw and Direct3D8 cooperatively.  Since d3d8to9 converts Direct3D8  into Direct3D9, I wonder if Microsoft changed the way DirectDraw and Direct3D9 work together cooperatively?

Note: I also found another game that uses DirectDraw and Direct3D8 cooperatively called [Port Royale](https://www.gog.com/game/port_royale).  This game also has issues on Windows 10, though the issues are different.

**Comment by elishacloud** [comment]
It looks like this issue is worse with the Windows 10 [April Update](https://support.microsoft.com/en-us/help/4028685/windows-10-get-the-update).  Now it's not only the intro video that shows a black screen, the other videos in the game also only show a black screen.

**Comment by elishacloud** [comment]
Yes, I tried "Disable fullscreen optimizations" as well as several other compatibility settings.  They have no effect.

**Comment by elishacloud** [comment]
Yes, this is not a compatibility issue.  This seems like an issue with how Windows 10 handles Direct3D8 differently than Direct3D9.

**Comment by elishacloud** [comment]
I found out that if I set the game to [Windowed mode](https://github.com/elishacloud/dxwrapper/commit/8615d33990c4b429bf96728847d287099421e234) then the videos play just fine with d3d8to9.

Apparently I just need to set:

```c++
pPresentationParameters->Windowed = true;
```

When creating the Direct3D device for the videos to play.

**Comment by Radio30** [comment]
Thanks @elishacloud ! That got the videos working but it seems to disable re-shade, is it the same on your end?

**Comment by elishacloud** [comment]
Here is an updated patch.  This one will no longer bypass ReShade:  [d3d8.zip](https://github.com/crosire/d3d8to9/files/2311092/d3d8.zip)

However it appears that ReShade has a bug.  ReShade crash for me with _Indiana Jones and the Emperor's Tomb_ even when using the released version of d3d8to9.  You may want to open a bug in [ReShade](https://reshade.me/forum) for this.

Note: the only after market injector I could get to work with this game is [injectSMAA](http://mrhaandi.blogspot.com/p/injectsmaa.html).

**Comment by Radio30** [comment]
I've just tested your new patch and it works perfectly with reshade! I'm using the presets Magic DOF, HDR, Adaptive sharpen and SMAA and it's working great with no problems at all! I'm using Reshade 3.0.7.180

**Comment by elishacloud** [comment]
Interesting.  It could be the version of the game I am using.   I am using the [GOG](https://www.gog.com/game/indiana_jones_and_the_emperors_tomb) version.  I am glad it works for you!

BTW: I am not sure what a generic fix for this would be.  Switching all games to use windowed fullscreen is probably not the best solution.  Adding an option for this would work.  However up until now all fixes have been generic so no configuration has been needed.

**Comment by Radio30** [comment]
No worries, I'm also using the GOG version.
Also I've found for some games, one being blood II: the chosen, having quicktime player allowed the intro movies to play but i'm not sure how this is connected.

**Comment by elishacloud** [comment]
> having quicktime player allowed the intro movies to play but i'm not sure how this is connected.

That is different.  That is likely because QuickTime installs some [codec](https://en.wikipedia.org/wiki/Codec) that is needed for the game movies.

For _Indiana Jones and the Emperor's Tomb_ the issue could be related to some display format (similar to issue #42).  When in windowed mode it converts the format to the current display format.  However in fullscreen it tries to use the format directly.  I will need to troubleshoot this more when I have some time.

**Comment by Radio30** [comment]
Hi elishacloud,
I've been playing Indiana Jones and The Emperor's Tomb happily with videos and d3d8to9 and it had been beautiful. However the latest windows update has made it so only the audio plays from the videos again and has also broken some d3d8to9 games I had up and running, It's such a pain.

I don't know if this topic still interests you But your help on here has been amazing and you're the only person I can think of who might know what caused this to happen

**Comment by elishacloud** [comment]
@Radio30, I am not quite sure exactly why it is just showing a black screen.  I believe it is a bug in Windows 10 and can reproduce this with other games using other versions of DirectX.  However, I have found that running in fullscreen widowed mode solves the issue.

I had to make some changes to my windowed mode code because of the latest Windows 10 release.  Below is a patch that works for now.  There is still one issue with this patch where the game will hang if you click the mouse during the opening video.  I will fix this later, but for now you can use this as long as you don't click the mouse during the opening video.

This is using the [dxwrapper](https://github.com/elishacloud/dxwrapper) code since it already has support for fullscreen windowed mode. Simply unzip these files into the `GameData\bin` folder under the game's folder:  [d3d8.zip](https://github.com/crosire/d3d8to9/files/3450294/d3d8.zip)

**Comment by Radio30** [comment]
Thankyou for the reply! I was excited to try this fix and have Indy working as normal again but sadly after launching, the first 2 intro movies play and then it's CTD without touching the mouse, or anything else :(

**Comment by elishacloud** [comment]
I believe the game does not respond to certain notification messages which cause this issue.  Those messages only happen because the game is running in windowed mode.  I will probably need to handle those messages myself.

Ideally we would get this fixed in exclusive fullscreen mode, but I am not quite sure how to do that.  I believe this is a bug in Windows, similar to what is described in this thread [here](https://github.com/narzoul/DDrawCompat/issues/15#issuecomment-297823521).

**Comment by Radio30** [comment]
Ah, cheers for the link, I had a read.
I like to have quite a few old games running with reshade, some using d3d8to9 and it seems each new windows update is a potential threat to the stability of older games running in such a way.

**Comment by Radio30** [comment]
@elishacloud 
Hey, Have you had a chance to look into this any further? I've tried a few things just messing around like converting the bik files to a different format but that was probably stupid as the game is looking for .bik files lol.
I'm so close to having a perfect running indy, it's just those damn movies

**Comment by elishacloud** [comment]
I did take a few minutes to look at this issue.  I found several other games that also show a black screen during the intro video when they are converted to Direct3D9.

Here are the other games:
 * American Conquest
 * Konung 1
 * Konung 2
 * SimCopter

These games are using older versions of DirectX, so it's not just converting from Direct3D8 to Direct3D9 that has this issue.

What appears to be happening is that the videos are **_not_** being displayed via DirectX at all.  They are using GDI to directly to write to the window.  Apparently with the latest version of Direct3D9 on Windows 10 it will no longer show data written directly to the window while in exclusive fullscreen mode.  At least I have not figured out any way of showing data written directly to the window while in Direct3D9 exclusive fullscreen mode.

The solution I have so far is to run these games in windowed fullscreen mode.  That used to work with this game, but with the latest version of Windows 10 a new issue has developed where forcing the game into windowed mode causes the game window to freeze/hang.  I believe the issue is that the game cannot handle some window event, which causes the window to hang.

A better way might be to hook GDI and convert the data to Direct3D9.  However, that is outside the scope of this project.

Anyways, I was able to build a slightly better way to enable windowed fullscreen mode.  This one should work as long as you don't click the mouse button during any of the intro videos.  At least it worked for me.  Later I will filter out mouse clicks during the intro videos, but this one should work for now.

Here is the update: [dxwrapper.zip](https://github.com/crosire/d3d8to9/files/3554763/dxwrapper.zip)

**Comment by mirh** [comment]
Isn't *direct forceful rendering to window* the same thing @CookiePLMonster faced with SilentPatchGF?

**Comment by Radio30** [comment]
@elishacloud 
Wow!! You've done it again! 👍 Thankyou so much for posting this fix, works perfectly, I'm very happy!!

You're a proper genius!

**Comment by CookiePLMonster** [comment]
> Isn't _direct forceful rendering to window_ the same thing @CookiePLMonster faced with SilentPatchGF?

Seemingly not, as this doesn't seem to involve DirectDraw. It can be similar, however.

**Comment by elishacloud** [comment]
@CookiePLMonster, you can ignore the DirectDraw calls.  They seem useless.  I believe they are just used to get video memory and set cooperative mode.  If you block all the DirectDraw functions the game continues to work correctly (even the intro videos).

How did you solve the issue with SilentPatchGF?

**Comment by CookiePLMonster** [comment]
> How did you solve the issue with SilentPatchGF?

I ended up leveraging my RenderWare knowledge and writing a game specific hook for it as described in detail here:
https://cookieplmonster.github.io/2018/05/18/fixing-the-godfather/

**Comment by elishacloud** [comment]
I see.  Nice work on the fix!  Unfortunately that appears to be a completely different issue.  I don't think that will help with this issue.

BTW: I am working on a generic DirectDraw -> D3D9 wrapper and I plan to support overlays.  I will do this by using an emulated surface.  The game will write to the emulated surface (probably created from a DC) then I will copy the data from the emulated surface to the real D3D9 surface.  I have already done this with 24-bit surfaces, which are supported in DirectDraw but not in D3D9.

**Comment by elishacloud** [comment]
As I mentioned earlier, the issue here happens because the game uses GDI for the videos and DirectX for the game play.  However, in DirectX 9 when you are using exclusive fullscreen mode all GDI turns black.  So far the only way I know of to solve this is to put the game into fullscreen windowed mode, rather than exclusive fullscreen mode.  Then GDI and DirectX 9 can play nicely together.  But with this particular game if you force it into fullscreen windowed mode it causes the game to hang, which I may be able to fix this later.

However, just recently [dxvk](https://github.com/doitsujin/dxvk) added support for DirectX 9 conversion to Vulkan.  Vulkan is able to play nicely with GDI even in exclusive fullscreen mode.  So you can use d3d8to9 to convert the game to DirectX 9 and then dxvk to convert the game to Vulkan.  This should solve the black screen issue with the in-game videos.

**Comment by mirh** [comment]
> Vulkan is able to play nicely with GDI even in exclusive fullscreen mode

Couldn't that be the case because it leverages some DXGI/DWM voodoo?

**Comment by elishacloud** [comment]
Actually I tested dxvk with this game and it has the same issues.  It works with other games that have these issues but not this one.

I did note that dgVoodoo2's DirectX8 dll does work with this game.  So far it is the only one I know that works.

**Comment by mirh** [comment]
Supposedly, he's [hooking](https://www.vogons.org/viewtopic.php?f=59&t=51790&start=940#p753478) some GDI function and forcing its way out. Not sure if that's anything specific to the MCIAVI(?) library or it could be generalized.

**Comment by elishacloud** [comment]
Yes, hooking GDI is the same thing that [DDrawCompat](https://github.com/narzoul/DDrawCompat/blob/6aaf12b0a45bb01881b08a4f04e8808869fbe28a/DDrawCompat/Dll/DllMain.cpp#L73-L74) does.  I also mentioned GDI hooking earlier in [this thread](https://github.com/crosire/d3d8to9/issues/61#issuecomment-526104706).

I am considering adding GDI hooking to my [dxwrapper](https://github.com/elishacloud/dxwrapper) project.  However, for most other games I tested, simply setting the game to fullscreen windowed mode solves the issue.

In fact, even for _Indiana Jones and the Emperor's Tomb_ fullscreen windowed mode allows the video to be displayed.  The issue is that the game does not properly handle some hwnd events during the videos causing the game to hang when setting it to fullscreen windowed mode.  These events don't exist in exclusive fullscreen mode so the game never needed to handle them.  However, if you wait until video is completed then the game will recover and become playable again.

I am considering hooking the wndproc like what is done in [diablo-ddrawwrapper](https://github.com/strangebytes/diablo-ddrawwrapper/blob/eb7cbab8462b5d517ab167e2258d64a608b87587/Project/ddrawwrapper/DirectDrawWrapper.cpp#L494-L499) and [cnc-ddraw](https://github.com/CnCNet/cnc-ddraw/blob/master/src/main.c#L1658).  Then I can simply filter out the events that the game does not handle. That would be much easier than writing a whole GDI wrapper.

**Comment by mirh** [comment]
And what about surface sharing with GDI interop instead?

**Comment by elishacloud** [comment]
ddraw and Direct3D8 seem to handle GDI interop ok.  But I have no idea how to do GDI interop with Direct3D9.  I did add a feature recently to dxwrapper to allow Direct3D9 to [Blt to GDI](https://github.com/elishacloud/dxwrapper/blob/b0ad9aea3855c98fa8cf930c8cb71dfd603be3a2/ddraw/IDirectDrawSurfaceX.cpp#L3173-L3188) rather than Direct3D9.  This helps resolve [an issue where ddraw mixes GDI and DirectX on the same screen](https://github.com/elishacloud/dxwrapper/issues/45#issuecomment-546663496) at the same time.

**Comment by mirh** [comment]
D3DXLoadSurfaceFromMemory/IDirect3DSurface9::GetDC maybe?
There's a [whole subway](https://web.archive.org/web/20100923230218/http://blogs.msdn.com/b/directx/archive/2009/08/26/directx-componentization-and-interoperability.aspx) of interoperability it seems.

**Comment by elishacloud** [comment]
[GetDC](https://docs.microsoft.com/en-us/windows/win32/api/d3d9helper/nf-d3d9helper-idirect3dsurface9-getdc) is used to allow a game to use GDI functions directly on DirectX.  It was added in [version 1](https://github.com/elishacloud/dxwrapper/blob/b0ad9aea3855c98fa8cf930c8cb71dfd603be3a2/ddraw/Versions/IDirectDrawSurface.cpp#L104-L107) and worked all the way to [version 9](https://github.com/elishacloud/dxwrapper/blob/b0ad9aea3855c98fa8cf930c8cb71dfd603be3a2/d3d9/IDirect3DSurface9.cpp#L202-L207).  DirectX 10 uses a different way of handling GDI.

The trouble is that some games (like this one) call the GDI functions directly rather than using DirectX for GDI.  To use this, I would have to hook the GDI functions and redirect them to DirectX using `GetDC()`.  But, that is probably the right way to do it.

**Comment by mirh** [comment]
You might also get fun with D3DSWAPEFFECT_FLIPEX/IDXGISurface1::GetDC/ID2D1GdiInteropRenderTarget for moar speed, if supported by the hardware :p

**Comment by elishacloud** [comment]
Ok, I put a fix for this in dxwrapper.  I just filtered the window events that were causing it to hang.  This should be safe for any game.

Here is the update: [dxwrapper.zip](https://github.com/crosire/d3d8to9/files/4006334/dxwrapper.zip)

Just unzip this into the `GameData\bin` folder of _Indiana Jones and the Emperor's Tomb_.

**Comment by elishacloud** [comment]
I made a few changes to prevent possible sync issues.  This one also has anti-aliasing enabled.

Here is the update: [dxwrapper.zip](https://github.com/crosire/d3d8to9/files/4007653/dxwrapper.zip)

_Note: I just realized that it still hangs sometimes if Task Manager is open.  Make sure to close the Task Manager before running._

**Comment by BehzadScorpion** [comment]
I went to ChooseRenderer.exe and changed the renderer to "Software MMX: FullScreen..." and it worked! I also changed the compatibility setting to win98 but I doubt it had anything to do with that cause when I did that without switching renderer it didn't work

**Comment by SwiftyOne1** [comment]
Hello, Sorry I'm just a user...I have been following this thread as I used the d3d8.dll in Nascar2003 season, it seemed to work although I could not see a lot of differents. I lost d3d8.dll and came here to find the latest build I tried quite a few d3d8.dll but all would crash the game so I didn't try any more until I tried dxwrapper posted by elishacloud over a year ago, I put all three files into the root directory of Nascar 2003 season, the game works fine and no crashing is there anything else that I could to make the game look better? Maybe that would need some work in the dxwrapper.ini file? Thank you for your work and any help you could give an end-user :)

**Comment by elishacloud** [comment]
The issue with _Indiana Jones and the Emporer's Tomb_ showing a black screen during the in-game videos is because it uses Bink to write directly to the GDI. However, in later versions of Windows 10 then Direct3D9 will over write anything written directly to GDI when in exclusive fullscreen mode.  The solution for the black screen is to switch to using windowed fullscreen mode.

As far as the game hanging when using windowed fullscreen mode, this is a bug with the version of Bink included in the game.  The version included is 1.5.19.  The issue is fixed in Bink 1.5.21.

Here is a package that includes all the fixes for this issue, including the updated version of Bink: [d3d8.zip](https://github.com/crosire/d3d8to9/files/6247293/d3d8.zip)

**Comment by elishacloud** [comment]
I added a fix for this in dxwrapper so that even if you use the broken Bink 1.5.19 it will still not hang.

Here is the updated fix:  [d3d8.zip](https://github.com/crosire/d3d8to9/files/9874465/d3d8.zip)

**Comment by elishacloud** [comment]
I found a real fix for this issue by calling [`Direct3D9SetSwapEffectUpgradeShim()`](https://github.com/crosire/reshade/commit/3fe0b050706fb9f3510ed48d619cad71f7cb28f2#diff-74772e50e2921e0bf69f0470e81d072d30c071b6e4b30af8cfd6cda58cf249eeR35) with parameter of `0`.  This changes the present mode to "Hardware: Legacy Flip", which is the same mode that DirectDraw and Direct3D8 is using.  For more details on the issue see comments [here](https://github.com/narzoul/DDrawCompat/issues/269#issuecomment-1843829944).

Here is the fix for this in dxwrapper:  [dxwrapper.zip](https://github.com/crosire/d3d8to9/files/13607620/dxwrapper.zip)

@crosire, not sure if we want to put a fix in for this.  It does effect some other games.  Basically any Direct3D8 game that uses GDI to display anything on the screen will break without this fix.  Older games that used Bink often used GDI for videos, as this game does.  [Here](https://github.com/elishacloud/dxwrapper/commit/16364ecbad5f7c807f1df11058605d651df9310f) is the code fix I put into dxwrapper for this issue.

**Comment by mirh** [comment]
A quick and dirty search @WinDLLsExports shows that while `SwapEffectUpgradesShim` is already shipping in W10 dlls, `SwapEffectUpgradeEnable` (and `DX_DATABASE_QUERY_TYPE_SWAPEFFECTUPGRADE_ENABLE`) is only present from W11. 

Unsure if this means that in older Windows you can only control it through *GameConfigStore* then (this is the only thing I could remotely recognize from a very blind run in ghidra), or if that couldn't just be dead code that microsoft is sharing between different versions for lazy reasons.

EDIT: there's also `SwapEffectUpgradeCache` in `HKCU\Software\Microsoft\DirectX\GraphicsSettings` for some reason

**Comment by mirh** [comment]
It has "shim" in the name, but after entering a colossal rabbit hole I just found out it's not a normal one implemented in the general application compatibility layers..

First of all, `SwapEffectUpgradesShim` inside of d3d9 and dxgi was added in W10 1709 (which was also about "peak FSO", [before](https://www.pcgamingwiki.com/wiki/Windows#Fullscreen_optimizations) they scaled back on the amount of automatic upgrades). Maybe it was a new knob to toggle what they had already been doing for a while, maybe they refined or covered only some specific cases back then, but nonetheless today it is what it is (W10 1803 further added `SwapEffectTransitionShim` but idk). 

Finally Windows 11 added `WindowedSwapEffectUpgrade` (which of course controls [this](https://www.elevenforum.com/t/turn-on-or-off-optimizations-for-windowed-games-in-windows-11.4943/)) and this "DirectX Database query" thing that I spotted. After some creative search, I guessed that refers to *DirectXApps.sdb*, *DirectXApps_FOD.sdb* or *DirectXApps_OEM.sdb* which should have been introduced from W10 1903 onwards (and which only the god knows in functioning: presumably that's associated with a *directxdatabaseupdater.exe* scheduled task and starting in W11 with *directxdatabasehelper.dll* too).
EDIT: the FOD variant was added in [20H2](https://learn.microsoft.com/en-us/windows-hardware/manufacture/desktop/features-on-demand-non-language-fod#directx-configuration-database) and removed(?) by [22H2](https://www.deploymentresearch.com/new-and-removed-built-in-apps-in-win11-22h2/). I also assume that the main one is handled by a [onesettings](https://patchmypc.com/blog/how-controlled-feature-rollout-actually-works/) feature `"`[`Ctac`](https://learn.microsoft.com/en-us/windows/privacy/basic-level-windows-diagnostic-events-and-fields-1809#microsoftwindowsremediationapplicable)`AppId":"DXDB"` (the [queryurl](https://settings.data.microsoft.com/settings/v3.0/sigma/DirectXDatabase) somehow returns nothing atm, but the last file is [this](https://fs.microsoft.com/fs/directxdatabase/1.8.8/DirectXApps.lzma)) which should be related to dword `HybridDeviceApplicableForDxDbGpuPreferences`.

Last but not least, after opening them with SDB explorer (because the ACT tools don't seem to parse anything besides the entries names) I found they are used to set some custom value for settings like `GpuPreference`, `AutoHDREnable`, `SwapEffectUpgradeEnable`, `DynamicSwitch` and `DXGIEffects`. The absolute lack of recognition of even just the existence of all of this kinda baffles me.

**Comment by elishacloud** [comment]
Thanks.  A lot of good data.

> Finally Windows 11 added `WindowedSwapEffectUpgrade` (which of course controls [this](https://www.elevenforum.com/t/turn-on-or-off-optimizations-for-windowed-games-in-windows-11.4943/))

I just tested the steps [here](https://www.elevenforum.com/t/turn-on-or-off-optimizations-for-windowed-games-in-windows-11.4943/) and enabled "flip presentation model" and it did not fix the issue here with a black screen.  I am not quite sure what it does, but it is not a fix for this issue.

> I found they are used to set some default value for settings like `GpuPreference`, `AutoHDREnable` and `SwapEffectUpgradeEnable`.

I installed the latest Compatibility Manager for Windows 11 and I was unable to find any Compatibility Fix for `SwapEffectUpgradeEnable`.  If you know a way to enable this outside of calling the API I listed above I would be interested in it.

**Comment by mirh** [comment]
> I just tested the steps [here](https://www.elevenforum.com/t/turn-on-or-off-optimizations-for-windowed-games-in-windows-11.4943/) and enabled "flip presentation model" and it did not fix the issue here with a black screen.

If calling the d3d9 upgrade shim with parameter 0 did it, then presumably what you actually want was the opposite.

> I installed the latest Compatibility Manager for Windows 11 and I was unable to find any Compatibility Fix for SwapEffectUpgradeEnable.

I don't think *Compatadmin* is actually supposed to handle the directx database, because its strings aren't even present in the usual apphelp places (it's only in d3d9.dll and dxgi.dll.. and well, somehow opengl32.dll and ahcache.sys too but I'm digressing). Use [SDB explorer](https://ericzimmerman.github.io).

> If you know a way to enable this outside of calling the API I listed above I would be interested in it.

Supposedly there is an [API](https://learn.microsoft.com/en-us/windows/win32/devnotes/application-compatibility-database)... but I'm not sure how useful it may be.

----
> Unsure if this means that in older Windows you can only control it through GameConfigStore then

Well, well, well, well... *Well*.
After a week of banging my head I did find such perverse mechanism, and boy it's some work of the devil.

So.. in the beginning there is **KnownGameList.bin** (which is present in `C:\Windows\bcastdvr\`, `%LOCALAPPDATA%\Microsoft\GameDVR\` and wherever `Microsoft.XboxGameOverlay_8wekyb3d8bbwe` is installed).
You may think this is just some innocent "quality of life" list of executable names just to help the DVR service heuristics identify games, but you couldn't be wronger and it couldn't be more wickedly [entrenched](https://www.rockpapershotgun.com/windows-10-game-mode-explained) into the system.
I only eventually realized this because it was the only place on earth that was somehow including bangers like "pcsx2-r4600.exe" (a name that I figured could get every dx11 application to be upgraded to flip presentation on my W10 22H2 laptop).

And where does it enter the picture you'll ask? Well, it's actually loaded right straightly by \*explorer.exe\* (courtesy of *twinui.dll*) every time it starts. And not just that then, if you delete/rename the file after taking ownership it will always try to restore it from the copy that ships in the aforementioned UWP package (note that the version in *%windir%* can also get bumped up by windows [updates](https://support.microsoft.com/en-us/topic/august-13-2024-kb5041580-os-builds-19044-4780-and-19045-4780-2ef55b0d-bb01-41c8-8629-4146929792ad#ID0EPBD), while there's a whole [task](https://strontic.github.io/xcyclopedia/library/clsid_E07647F7-AED2-48D9-9720-939BC24A8A3C.html) run with the DiagTrack/wosc service to [refresh](https://github.com/meetrevision/revision-tool/blob/2.0.3/common/lib/src/services/miscellaneous_service.dart#L167) the appdata one every few hours).
But even after killing the desktop environment (and restarting both dwm and the intel driver) it still wasn't enough. 

AFAICT, it's not even that these inject into applications directly.. They (or at least something else along the line that my monitoring missed) just set some unknown knobs into `HKCU\System\GameConfigStore`. 
That will then be read on d3d initialization by svchost (more specifically the BrokerInfrastructure/psmsrv service, which loads PsmServiceExtHost.dll and ResourcePolicyServer.dll) which at last must be what the graphics subsystem dlls comunicate with.
I still haven't figured the GUIDs and the parent/children logic (e.g. even if you touch the path identifiers they are getting regenerated on the next launch, but if you delete the whole key nothing gets back) but long story short at the end of this madness.. if I set DWORD "Type" from 1 to 0 I'm finally getting bitblt again.
Some people [could](https://forums.blurbusters.com/viewtopic.php?t=13067) [swear](https://forums.guru3d.com/threads/another-new-way-to-force-exclusive-full-screen.447515/page-2#post-6124578) changing the "Flags" value worked too (it's obviously a bitmask) but now I stopped my testing both due to tiredness.. and also because presentmon broke down.

**TL;DR: some Windows quirk can enable the game bar, or at least some of its precursors like flip mode, for pretty much thousands of "known" game executable names.** It's really mindblowing how [little](https://old.reddit.com/r/Windows10/comments/c5ahfc/bug_dx9dx11_games_do_not_enter_independent_flip/es3gcdk/?context=1) it was ever acknowledged.
(and precise details are still evading me, for instance on my desktop with the same W10 22H2 this didn't seem to happen)

p.s. something I also didn't test was checking whether analytic ETW couldn't be [logging](https://github.com/repnz/tracelogging-providers/blob/master/Win10_18363/tlg/d3d9.tlg#L237) these behaviours with events `gameConfigStoreEnabled_in`, `gameConfigStoreOverrideEnabled_in` or `swapEffectUpgradesShimEnabled_in` 
p.p.s. `adapterSupportsMaximizedWindowedModeForGameDVR_in` makes me remember about DXGIAdapterCache (which seems a close brother of DirectXDatabaseUpdater) too, also a mystery

**Comment by elishacloud** [comment]
Interesting.  I tried setting the registry keys under `HKCU\System\GameConfigStore` for this game but it did not help get the video to display.  The videos are using GDI to present to the screen.  I found the fix for this by calling an API, see [here](https://github.com/crosire/d3d8to9/issues/61#issuecomment-1846321943) for details.

**Comment by mirh** [comment]
I mean.. sure, I guess that's likely the direct way to do it (although I wished I could find the code or the caller in disassemblies).
But first I'd have tried to rename `indy.exe` just to confirm it's due to the allowlist to begin with. Everything else, I wouldn't really trust given the amount of absolute crazy nested tomfoolery involved.

---
I don't feel like bothering with further strings dumps here, though after some timid procmon'ing I still found much to wonder.
Putting aside that `gameconfighelper.dll` comes up too in the overall equation (and somehow not the one in the system32 folder but the one part of `Microsoft.GamingServices_8wekyb3d8bbwe`, which should be among the stuff pinged by *twinui.dll*) there are lots of very sus registry values.

First of all, children entries are not only queried for the few entries that are always present but also for `UserOverrideMask`, `AGCProfile`, `AGGProfile`, `AGCpuAlloc`, `Arguments`, `ProcessDependencies`, `WorkingDirectory` and `ExeParentDirectory` which I am scared to even think how they'd be used (even though I'll grant I could find the second last for BG3 and it's an innocent "bin").

Secondly, it appears there's something very strange going on in `HKCU\System\GameConfigStore` itself. On my system I have a DWORD named `GameDVR_FSEBehaviorMode`, which never gets checked. Conversely multiple times on d3d init the not-seen-elsewhere `GameDVR_FSEBehavior` is always poked (there's also a `Win32_GameModeUserRelatedProcesses` query tbh but go figure).

Last but absolutely not least, there is a `gameConfigStoreManagement` DWORD in 

    HKLM\SOFTWARE\Microsoft\SecurityManager\AdminCapabilities

Untested, but hopefully it could do something with tests in this regard without needing to [uproot](https://forums.guru3d.com/threads/how-to-disable-fso-in-w10.435057/) [everything](https://forums.mydigitallife.net/threads/creators-updates-game-functions-are-sinister-and-non-removable.74313/page-3).

**Comment by mirh** [comment]
https://pastebin.com/1tp0pA1j
FWIW the GameConfigStore format was recently figured.
https://github.com/SpecialKO/SpecialK/blob/main/src/wgc/wgc_registrar.cpp
```
the childside parent value is just LocalMachine-scoped DPAPI protected UTF-16LE of the lowercased basename with extension.
you can verify it yourself by taking any existing child parent blob and decrypting it with CryptUnprotectData/ProtectedData.Unprotect, it will yield the lowercase .exe filename.
protect that same UTF-16LE string again with CryptProtectData/ProtectedData.Protect using CRYPTPROTECT_LOCAL_MACHINE (0x4) and you will get a different raw blob that still decrypts to the same name.
```

> a name that I figured could get every dx11 application to be upgraded to flip presentation on my W10 22H2 laptop

As for the root of my puzzling, I still haven't figured it out after two years. Officially fully updated W10 shouldn't have any mechanism to upgrade the swap effect, yet on my 6500U (HD 520) it happens with the KGL whitelisted names. Conversely on my desktop 9600K (UHD 630) it *expectedly* doesn't, despite the latter even having slightly better MPO/PanelFitter caps and WDDM model.

**Comment by mirh** [comment]
Ok I finally got to the bottom of this after years.
First of all, I had completely [misunderstood](https://github.com/shoober420/windows11-scripts/blob/main/EnableFullscreenOptimizations.bat) how the *GameConfigStore* key works (please take the provided link with a grain of salt because not only its supposed sources are awfully trash and seem AI, but this is also the first time I ever heard of "Enhanced Fullscreen Exclusive").
For most of those knobs 0 keeps ***enabled*** FSO, while 1 is the override.
Setting `GameDVR_DXGIHonorFSEWindowsCompatible` to 1 is what finally made my laptop igp become "normal" again.

But still, the question remained of why this made a difference that anywhere else didn't happen even with the variable set to 0.
Even when I straight-up formatted a fresh disk with a clean copy of 22H2 19045 and shifted it between pcs, crazily enough the literal same installation with the literal [same](https://www.intel.com/content/www/us/en/download/762755/intel-6th-gen-processor-graphics-windows.html) driver had the same pattern persist.
I even tested a 6600K (HD 530) and that behaved like my other skylake system, ruling out that it could be a mobile platform quirk.
And I was starting to feel like a cucko when I realized that even MPOs were secondary to the issue, as when I [forced](https://www.pcgamingwiki.com/wiki/Topic:Ziyehx3oib42e33s) them off from the driver nothing changed (not that `Composed: Flip` required them tbh, but still you would guess that's the end goal of forcing this everywhere?).

Then finally I tried to install older drivers on my coffee lake desktop, as WDDM version was the [only](https://community.intel.com/t5/Graphics/WDDM-2-2-support-on-skylake/td-p/527058) differing variable that I had yet to manipulate (if somebody knows how to force which WDDM version to use from the OS side, please say).
Idk how much fucking lucky I was to find a driver in the windows catalog from 6 months before my cpu was even released (and most notably one month before W10 1803).. but long story short that did the trick. With a WDDM 2.3 driver even my newer system was getting forced flip model (in fact I was even getting `Independent Flip` here, probably due to the aforementioned slightly better MPO features).
I even installed older drivers for my 950M (alas my 2080 was too new for <2018 drivers) and I could confirm this isn't even intel specific either, going all the way down to replicating this with the original WDDM 2.0 drivers. 

**IF** you have WDDM older than 2.4 and **IF** you didn't touch the registry as mentioned above, you are getting forcedly upgraded to flip.
And it gives me ample satisfaction to note, that this thread was started right about in that period (at this point I even have the doubt if some setting ever actually changed between major Windows versions, or if it had always just been this to disable FSO after 1709).

p.s. unfortunately even my old laptop started to be *too new* for me to test WDDM 1.3 drivers, though I tried with an ancient radeon 5850 and FSO never enabled by itself (if somebody has a 2012~2014 gpu it would be nice to check if just switching drivers is enough to trigger this)

EDIT: obviously it might even be [some](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/d3dkmddi/ns-d3dkmddi-_dxgk_drivercaps) [other](https://learn.microsoft.com/en-us/windows-hardware/test/hlk/testref/05a86c55-9c50-4050-8932-30f951d72709) [correlated](https://github.com/shorthorn-project/One-Core-API-Source/blob/4.4.0/sdk/include/ddk/d3dkmdt.h#L2161) [bitflag](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/dispmprt/ns-dispmprt-_driver_initialization_data), but I have no further mean to investigate this without debugging interfaces
